docs/sbft-formal-model/distributed_system.s.dfy

```diff
@@ -0,0 +1,97 @@
+//#title DistributedSystem
+//#desc DO NOT EDIT THIS FILE! It's a trusted file that specifies how hosts interact
+//#desc with one another over the network.
+
+include "replica.i.dfy"
+include "client.i.dfy"
+include "cluster_config.s.dfy"
+include "messages.dfy"
+
+// Before we get here, caller must define a type Message that we'll
+// use to instantiate network.s.dfy.
+
+module DistributedSystem {
+  import opened HostIdentifiers
+  import opened Messages
+  import ClusterConfig
+  import Replica
+  import Client
+  import Network
+
+  datatype HostConstants = Replica(replicaConstants:Replica.Constants) | Client(clientConstants:Client.Constants)
+  datatype HostVariables = Replica(replicaVariables:Replica.Variables) | Client(clientVariables:Client.Variables)
+
+  datatype Constants = Constants(
+    hosts:seq<HostConstants>,
+    network:Network.Constants,
+    clusterConfig:ClusterConfig.Constants) {
+    
+    predicate WF() {
+      && clusterConfig.WF()
+      && clusterConfig.ClusterSize() == NumHosts()
+      && |hosts| == NumHosts()
+      && (forall id | clusterConfig.IsReplica(id) :: hosts[id] == HostConstants.Replica(Replica.Constants(id, clusterConfig)))
+      && (forall id | clusterConfig.IsClient(id) :: hosts[id] == HostConstants.Client(Client.Constants(id, clusterConfig)))
+    }
+  }
+
+  datatype Variables = Variables(
+    hosts:seq<HostVariables>,
+    network:Network.Variables<Message>) {
+    
+    predicate WF(c: Constants) {
+      && c.WF()
+      && |hosts| == |c.hosts|
+      && (forall id | c.clusterConfig.IsReplica(id) :: hosts[id].Replica? && hosts[id].replicaVariables.WF(c.hosts[id].replicaConstants))
+      && (forall id | c.clusterConfig.IsClient(id) :: hosts[id].Client? && hosts[id].clientVariables.WF(c.hosts[id].clientConstants))
+    }
+  }
+
+  predicate IsCommitted(c:Constants, v:Variables, view:nat, seqID:SequenceID) {
+    && v.WF(c)
+    && var ReplicasCommitted := set replicaID | && 0 <= replicaID < c.clusterConfig.N()
+                                                && Replica.QuorumOfPrepares(c.hosts[replicaID].replicaConstants,
+                                                                            v.hosts[replicaID].replicaVariables,
+                                                                            seqID);
+    //TODO: Implement check for hasPrePrepare
+    && |ReplicasCommitted| >= c.clusterConfig.ByzantineSafeQuorum()
+  }
+
+  predicate Init(c:Constants, v:Variables) {
+    && v.WF(c)
+    && (forall id | && c.clusterConfig.IsReplica(id) 
+                    :: Replica.Init(c.hosts[id].replicaConstants, v.hosts[id].replicaVariables))
+    && (forall id | && c.clusterConfig.IsClient(id)
+                    :: Client.Init(c.hosts[id].clientConstants, v.hosts[id].clientVariables))
+    && Network.Init(c.network, v.network)
+  }
+
+  // JayNF
+  datatype Step = Step(id:HostId, msgOps: Network.MessageOps<Message>)
```

Comment:
same as above: "Jay Normal Form (JNF)". You could also provide a link to an explanation of what it means.